### PR TITLE
docs(style): update style guide with hooks info

### PR DESCRIPTION
**Pull request by joshblack**
This PR updates our style guide with some hook info/best practices that have come up in the past couple of sprints that seemed helpful to codify.

Super curious what everyone thinks! 👀 

#### Changelog

**New**

**Changed**

- Update style guide with hook info/patterns/best practices

**Removed**

#### Testing / Reviewing

- Check out the added style guide sections and see if they make any sense 😅

### CODE REVIEW

**Comment by joshblack** [comment]
FYI @carbon-design-system/developers-system added a couple of sections to the style guide, would love to hear your thoughts! 👀

**Review by jnm2377 — APPROVED** [review]

**Comment by tw15egan** [comment]
This is really handy!

**Review by tw15egan — APPROVED** [review]

File: docs/style.md [hunk 1]
```diff
@@ -315,6 +318,123 @@ function MyComponent({ translateWithId: t = translateWithId }) {
 }
 ```
 
+#### Using `useCallback` and `useMemo`
+
+`useCallback` and `useMemo` can be incredibly useful tools in certain
+situations. In general, however, we try to avoid them unless one of the
+following conditions occur:
+
+- The identitiy of a function or object is required as a dependency in a
```
Comment (emyarod) [inline docs/style.md:327]:
```suggestion
- The identity of a function or object is required as a dependency in a
```

File: docs/style.md [hunk 2]
```diff
@@ -315,6 +318,123 @@ function MyComponent({ translateWithId: t = translateWithId }) {
...
+- We have observed performance issues due to allocations that can be reproduced
+  and resolved using these techniques
+
+This practice is to avoid introducing `useCallback` and `useMemo` prematurely,
+which can create extra work for our components to perform.
+
+A rule of thumb for this is to understand how frequently a dependency will
+update that is given to `useCallback` or `useMemo`. If a dependency is likely to
+update frequently, then React will have to perform comparisons and re-run
+callback to `useCallback` and `useMemo`. This would be slower than creating a
```
Comment (emyarod) [inline docs/style.md:338]:
```suggestion
callbacks to `useCallback` and `useMemo`. This would be slower than creating a
```

File: docs/style.md [hunk 3]
```diff
@@ -315,6 +318,123 @@ function MyComponent({ translateWithId: t = translateWithId }) {
...
+- The identitiy of a function or object is required as a dependency in a
+  dependency array
+- We have observed performance issues due to allocations that can be reproduced
+  and resolved using these techniques
+
+This practice is to avoid introducing `useCallback` and `useMemo` prematurely,
+which can create extra work for our components to perform.
+
+A rule of thumb for this is to understand how frequently a dependency will
+update that is given to `useCallback` or `useMemo`. If a dependency is likely to
```
Comment (emyarod) [inline docs/style.md:336]:
```suggestion
A rule of thumb for this is to understand how frequently a dependency that 
is given to `useCallback` or `useMemo` will update. If a dependency is likely to
```

**Pull request merged**